DOCSP-9557: Fundamentals > CRUD Operations > Read Operations > Sort Results (#45)
mongodb/docs-java: source/fundamentals/crud/sort.txt

insert this data into a collection, run the following operation:

.. code-block:: java

    collection.insertMany(Arrays.asList(
        new Document().append("_id", 1)
            .append("title", "The Brothers Karamazov").append("length", 824)
            .append("author", "Dostoyevsky"),
        new Document().append("_id", 2)
            .append("title", "Les Misérables").append("length", 1462).append("author", "Hugo"),
        new Document().append("_id", 3)
            .append("title", "Atlas Shrugged").append("length", 1088).append("author", "Rand"),
        new Document().append("_id", 4)
            .append("title", "Infinite Jest").append("length", 1104).append("author", "Wallace"),
        new Document().append("_id", 5)
            .append("title", "Cryptonomicon").append("length", 918).append("author", "Stephenson"),
        new Document().append("_id", 6)
            .append("title", "A Dance with Dragons").append("length", 1104)
            .append("author", "Martin")
    ));

Call the ``sort()`` function and specify the order with ``ascending()`` to ensure that the
operation returns books in alphabetical order by title:

.. code-block:: java
  :emphasize-lines: 5

>     import com.mongodb.client.model.Sorts;

Reviewer: I don't think this is enough to make `ascending` work below.  You either have to [statically import](https://docs.oracle.com/javase/8/docs/technotes/guides/language/static-import.html) `Sorts.ascending` or use `Sorts.ascending` in the code.

    // define a cursor that can iterate through the results of the query
    MongoCursor<Document> cursor = collection.find()
        .sort(ascending("title")).iterator();
    // print out items in alphabetical order
    try {
      while (cursor.hasNext()) {
        System.out.println(cursor.next());
      }
    // close the cursor
    } finally {
      cursor.close();
    }

This sample code prints the documents in ascending order on the ``title`` field as follows:

.. code-block:: java

    Document{{_id=6, title=A Dance with Dragons, length=1104, author=Martin}}
    Document{{_id=3, title=Atlas Shrugged, length=1088, author=Rand}}
    Document{{_id=5, title=Cryptonomicon, length=918, author=Stephenson}}
    Document{{_id=4, title=Infinite Jest, length=1104, author=Wallace}}
    Document{{_id=2, title=Les Misérables, length=1462, author=Hugo}}
    Document{{_id=1, title=The Brothers Karamazov, length=824, author=Dostoyevsky}}

Sometimes, the order of two or more documents is ambiguous using a
specified sort. Consider the case of sorting by length: in the above case, 

> both "A Dance with Dragons" and "Infinite Jest" have ``1104`` pages, so 

Reviewer: Nice examples!

the order in which ``sort()`` returns them is not guaranteed. To resolve ties in 
your sorted results in a repeatable way, add additional fields to the 
``Sorts.ascending()`` function call. 
Call the ``sort()`` method, specifying ``length`` as 
the first field by which to sort and the ``title`` field as the second field. In the event 
that the ``length`` field comparison results in a tie, the operation compares the tied results 
using the ``title`` field:

.. code-block:: java
  :emphasize-lines: 3
    
    // sort first by length, then title
    MongoCursor<Document> cursor = collection.find()

>         .sort(ascending("length", "title")).iterator();

Reviewer: You might want to show an example where one field is ascending and another is descending.

============
Sort Results
============

.. default-domain:: mongodb

Use ``sort`` to change the order in which read operations return
documents. The ``sort`` function tells MongoDB to order returned documents by the
values of one or more fields in a certain direction. To sort returned
documents by a field in ascending (lowest first) order, use the ``Sorts.ascending``
function. To sort in descending (greatest first) order instead, use ``Sorts.descending``.
If you do not specify a sort, MongoDB does not guarantee the order of
query results.

Follow the instructions in the examples below to insert data into
a collection and perform a sort on the results of a query.
Consider a collection containing documents that describe books. To
insert this data into a collection, run the following operation:

.. code-block:: java

   collection.insertMany([
        new Document().append("_id", 1)
            .append("title", "The Brothers Karamazov").append("length", 824).append("author", "Dostoyevksy"),
        new Document().append("_id", 2)
            .append("title", "Les Misérables").append("length", 1462).append("author", "Hugo"),
        new Document().append("_id", 3)

>             .append("title", "Atlas Shrugged).append("length", 1088).append("author", "Rand"),

Reviewer 2: ```suggestion
            .append("title", "Atlas Shrugged").append("length", 1088).append("author", "Rand"),
```
Missing a closing quote. Copying and pasting this makes IntelliJ very mad.

>             .append("title", "Atlas Shrugged).append("length", 1088).append("author", "Rand"),

Reviewer 3: ```suggestion
            .append("title", "Atlas Shrugged").append("length", 1088).append("author", "Rand"),
```
The syntax highlighting for this code block looked absolutely crazy to me, until I realized that there was a missing quote 🤭

>    collection.insertMany([
>         new Document().append("_id", 1)
>             .append("title", "The Brothers Karamazov").append("length", 824).append("author", "Dostoyevksy"),
>         new Document().append("_id", 2)
>             .append("title", "Les Misérables").append("length", 1462).append("author", "Hugo"),
>         new Document().append("_id", 3)
>             .append("title", "Atlas Shrugged).append("length", 1088).append("author", "Rand"),
>         new Document().append("_id", 4)
>             .append("title", "Infinite Jest").append("length", 1104).append("author", "Wallace"),
>         new Document().append("_id", 5)
>             .append("title", "Cryptonomicon").append("length", 918).append("author", "Stephenson"),
>         new Document().append("_id", 6)
>             .append("title", "A Dance with Dragons").append("length", 1104).append("author", "Martin"),
>    ]);

Reviewer 2: ```suggestion
        collection.insertMany(new ArrayList<Document>(Arrays.asList(
                new Document().append("_id", 1)
                        .append("title", "The Brothers Karamazov").append("length", 824).append("author", "Dostoyevksy"),
                new Document().append("_id", 2)
                        .append("title", "Les Misérables").append("length", 1462).append("author", "Hugo"),
                new Document().append("_id", 3)
                        .append("title", "Atlas Shrugged").append("length", 1088).append("author", "Rand"),
                new Document().append("_id", 4)
                        .append("title", "Infinite Jest").append("length", 1104).append("author", "Wallace"),
                new Document().append("_id", 5)
                        .append("title", "Cryptonomicon").append("length", 918).append("author", "Stephenson"),
                new Document().append("_id", 6)
                        .append("title", "A Dance with Dragons").append("length", 1104).append("author", "Martin")
        )));
```
I'm unsure this is valid Java syntax for array declarations? Also, I was unable to use this code to insert documents even with adding the missing quote. `insertMany` appears to want a List like item.

    // print out items in alphabetical order
    try {
      while (cursor.hasNext()) {
        System.out.println(cursor.next());
      }
    // close the cursor
    } finally {
      cursor.close();
    }

This operation returns the following documents when using ``sort()`` with a query 
on ``title`` and an ascending order:

.. code-block:: java

    Document{{_id=6, title=A Dance with Dragons, length=1104, author=Martin}}
    Document{{_id=3, title=Atlas Shrugged, length=1088, author=Rand}}
    Document{{_id=5, title=Cryptonomicon, length=918, author=Stephenson}}
    Document{{_id=4, title=Infinite Jest, length=1104, author=Wallace}}
    Document{{_id=2, title=Les Misérables, length=1462, author=Hugo}}
    Document{{_id=1, title=The Brothers Karamazov, length=824, author=Dostoyevsky}}

Sometimes, the order of two or more documents is ambiguous using a
specified sort. Consider the case of sorting by length: in the above case, 
both "A Dance with Dragons" and "Infinite Jest" have ``1104`` pages, so 
the order in which ``sort()`` returns them is not guaranteed. To resolve ties in 
your sorted results in a repeatable way, add additional parameters to the 

> ``Sorts.ascending()`` function call:

Reviewer 4: Make this parallel in structure to the prior sample description:

```suggestion
``Sorts.ascending()`` function call.

Call the `sort()``method, specifying the length as the first field by which to sort and the length field as the second field. In the event that the length field comparison results in a tie, the operation compares the tied results using the title field:
```

.. code-block:: java
  :emphasize-lines: 5
    
    import com.mongodb.client.model.Sorts;

    // sort first by length, then title
    MongoCursor<Document> cursor = collection.find()
        .sort(ascending("length", "title")).iterator();
    // print out items
    try {
      while (cursor.hasNext()) {
        System.out.println(cursor.next());
      }
    // close the cursor
    } finally {
      cursor.close();
    }

> With the addition of the ``title`` field to the ``sort()`` call, the
> operation sorts matching documents first by ``length`` and, in the event
> of a tie, then by ``title``. The sort operation compares matched document fields in
> the same order as they are specified in the function call. The ``sort()``
> operation returns the following ordering of documents when used on
> the documents matching the query, sorting "A Dance with Dragons" before 
> "Infinite Jest" for the two books with the same length:

Reviewer 4: ```suggestion
This sample code prints your query results in ascending order by length. When there are two documents with the same value, it resolves the tie using the title field you specified as the second parameter.
```

    Document{{_id=4, title=Infinite Jest, length=1104, author=Wallace}}
    Document{{_id=2, title=Les Misérables, length=1462, author=Hugo}}

You can also sort by multiple fields in different orders, using ``Sorts.orderBy()``. 
Consider this example where you first sort on ``length`` in an ascending order, and then on
``title`` in a descending order:

.. code-block:: java
  :emphasize-lines: 3
    
    // sort first by length ascending, then title descending
    MongoCursor<Document> cursor = collection.find()

>         .sort(orderBy(ascending("length"), descending("title")))

Reviewer: Should we have imports for this example, to be consistent?

Author: My thoughts for that example were to only show the code that has changed from the previous examples (so, just the sort call), but I can add the import if you think it would clarify things. I can also add a sentence in explaining that any omitted code stays constant from one example to the next.

Reviewer: Right.  But the previous example statically imports just `import static com.mongodb.client.model.Sorts.ascending` and not any other static method from `Sorts`.  One option is to change the import to a wildcard, like `import static com.mongodb.client.model.Sorts.*`.   

@ccho-mongodb do we do that elsewhere for builders?

Reviewer 4: We haven't documented builders yet, but I used explicit static method imports for a snippet in this data formats example:

https://github.com/mongodb/docs-java/blob/master/source/fundamentals/data-formats/document-data-format-pojo.txt#L169-L172

I prefer listing them individually over using "*" since it's more clear what package the method comes from, but am flexible.
